library/core/src/hint.rs

```diff
@@ -51,18 +51,54 @@ pub const unsafe fn unreachable_unchecked() -> ! {
     unsafe { intrinsics::unreachable() }
 }
 
-/// Emits a machine instruction hinting to the processor that it is running in busy-wait
-/// spin-loop ("spin lock").
+/// Emits a machine instruction to signal the processor that it is running in
+/// a busy-wait spin-loop ("spin lock").
+///
+/// Upon receiving the spin-loop signal the processor can optimize its behavior by,
+/// for example, saving power or switching hyper-threads.
+///
+/// This function is different from [`thread::yield_now`] which directly
+/// yields to the system's scheduler, whereas `spin_loop` does not interact
+/// with the operating system.
+///
+/// A common use case for `spin_loop` is implementing bounded optimistic
+/// spinning in a CAS loop in synchronization primitives. To avoid problems
+/// like priority inversion, it is strongly recommended that the spin loop is
+/// terminated after a finite amount of iterations and an appropriate blocking
+/// syscall is made.
+///
+/// **Note**: On platforms that do not support receiving spin-loop hints this
+/// function does not do anything at all.
+///
+/// # Examples
 ///
-/// For a discussion of different locking strategies and their trade-offs, see
-/// [`core::sync::atomic::spin_loop_hint`].
+/// ```
+/// use std::sync::atomic::{AtomicBool, Ordering};
+/// use std::sync::Arc;
+/// use std::{hint, thread};
+///
+/// // Baby not born yet
```

Comment:
Maybe we should just be totally abstract with this and not reference people in any way 🙂

Let's just reference this as "the value", so we say "the value is live", and "the value is not live". The actual source itself looks good to me.

Reply:
I don't think many people relate to applying a spin loop hint in the 'real life' situation this example sketches. This example makes me personally quite uncomfortable.

Reply:
@pickfire I was thinking of something just like this:

```rust
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::{hint, thread};

// A shared atomic value that threads will use to coordinate
let live = Arc::new(AtomicBool::new(false));

// In a background thread we'll eventually set the value
let bg_work = {
    let live = live.clone();
    thread::spawn(move || {
        // Do some work, then make the value live
        do_some_work();
        live_clone.store(true, Ordering::Release);
    })
};

// Back on our current thread, we wait for the value to be set
while live.load(Ordering::Acquire) {
    // The spin loop is a hint to the CPU that we're waiting, but probably not for very long
    hint::spin_loop();
}

// The value is now set
do_some_work();
bg_work.join()?;
```

It's the same shape, but all we're trying to do is offer a little context for the function in the form of some surrounding source code, rather than try explain the concept itself.

Reply:
Thanks, I didn't get to give a good example. Lazy to keep up to this so I just took yours.